filters/CoordOperationFilter.cpp

```diff
@@ -0,0 +1,143 @@
+*       written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+* COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+* OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+* AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+* OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+* OF SUCH DAMAGE.
+****************************************************************************/
+
+#include "CoordOperationFilter.hpp"
+
+#include <gdal_version.h>
+#if GDAL_VERSION_MAJOR >= 3
+
+#include <pdal/PointView.hpp>
+#include <pdal/private/SrsTransform.hpp>
+#include <pdal/util/ProgramArgs.hpp>
+
+#include <ogr_spatialref.h>
+
+//#include <memory>
+
+namespace pdal
+{
+
+static StaticPluginInfo const s_info
+{
+    "filters.coordoperation",
+    "Transform coordinates using Proj pipeline string, WKT2 coordinate operations or URN definition",
+    "http://pdal.io/stages/filters.coordoperation.html"
+};
+
+CREATE_STATIC_STAGE(CoordOperationFilter, s_info)
+
+std::string CoordOperationFilter::getName() const { return s_info.name; }
+
+CoordOperationFilter::CoordOperationFilter()
+{}
+
+
+CoordOperationFilter::~CoordOperationFilter()
+{}
+
+
+void CoordOperationFilter::addArgs(ProgramArgs& args)
+{
+    args.add("a_srs", "Output spatial reference", m_aSRS);
+    args.add("reverse_transfo", "Wether the coordinate operation should be evaluated in the reverse path",
+             m_reverseTransfo, false);
+    args.add("coord_op", "Coordinate operation (Proj pipeline or WKT2 string or urn definition)",
+             m_coordOperation).setPositional();
+}
+
+
+void CoordOperationFilter::initialize()
+{
+    setSpatialReference(m_aSRS);
+    createTransform(m_coordOperation, m_reverseTransfo);
+}
+
+
+void CoordOperationFilter::createTransform(const std::string coordOperation, bool reverseTransfo)
+{
+    m_coordTransform.reset(new CoordTransform(coordOperation, reverseTransfo));
+}
+
+
+PointViewSet CoordOperationFilter::run(PointViewPtr view)
+{
+    PointViewSet viewSet;
+    PointViewPtr outView = view->makeNew();
+
+    PointRef point(*view, 0);
+    for (PointId id = 0; id < view->size(); ++id)
+    {
+        point.setPointId(id);
+        if (processOne(point))
+            outView->appendPoint(*view, id);
+    }
+
+    viewSet.insert(outView);
+    return viewSet;
+}
+
+
+bool CoordOperationFilter::processOne(PointRef& point)
+{
+    double x(point.getFieldAs<double>(Dimension::Id::X));
```

Comment:
Should we allow the user to map other dimensions to XYZ? I can see scenarios where data is modified in a pipeline and copied to other dimensions.

Reply:
I'm not familiar with the mechanism that allows data to be copied to other dimensions. In the issue discussion, you told me that the transformation filter has this capability, but watching to the code I don't see clearly where in the code this is enabled. 
I'm new to PDAL development so maybe I've missed something.

Reply:
What I would propose is adding options that allow users to map "X=>OtherX", and upon `initialize`, the filter caches a reference to the OtherX dimension and uses it as X (same for Y and Z). That way if there were other dimensions in the PointView that were different kinds of XYZ, the user wouldn't have to `filters.ferry` data into XYZ and destroy their original values (important when constructing transformations).

I also wonder if we shouldn't allow output of the new XYZ coordinates into **new** XYZ dimensions to preserve the old coordinates too, again for the same reasons as above.

Reply:
I prefer ferry.